experimental typesafety phase 1 (#12019)
remix-run/react-router: decisions/0012-type-inference.md

   You could completely alter the URL path for a route module, and TypeScript would be none the wiser.

2. **Type inference within a route**

   Some types are defined within a route module but need to be inferred across exports.

   For example, loader data is defined by the return type of `loader` but needs to be accessed within the `default` component export:

   ```ts
   export function loader() {
     // define here 👇
     return { planet: "world" };
   }

   export default function Component() {
     // access here 👇
     const data = useLoaderData<typeof loader>();
   }
   ```

   Unlike the `useParams` generic, this isn't just a type cast.
   The `useLoaderData` generic ensures that types account for serialization across the network.
   However, it still requires you to type of `typeof loader` every time.

   Not only that, but complex routes get very tricky to type correctly.

   > What if the route also has a `clientLoader`?
   > Did you remember to change the generic to `typeof clientLoader`?
   > But in the initial SSR render, the `clientLoader` doesn't run, so really you'd need `typeof loader | typeof clientLoader`.

>    > Unless you also set `clientLoader.hydrate = true` _AND_ provided a `HydrateFallback`. Then `clientLoader` always runs so you'd want just `typeof clientLoader`

Reviewer: Came here to say this ☝️

>    > Unless you also set `clientLoader.hydrate = true` _AND_ provided a `HydrateFallback`. Then `clientLoader` always runs so you'd want just `typeof clientLoader`

Author: Changed it to a small explanation plus truth table

>    That `useLoaderData` generic starts to feel a lot like doing your taxes: there's only one right answer, Remix knows what it is, but you're going to get quizzed on it anyway.

Reviewer 2: I can confirm that in Brazil it's exactly like that too

While React Router will default to programmatic routing, it can easily be configured for file-based routing.
That means that sometimes route URLs will only be represented as file paths.
Unfortunately, TypeScript cannot use the filesystem as part of its type inference nor type checking.
The only tenable way to infer types based on file paths is through code generation.

To that end, React Router will generate types for each route module into a special, gitignored `.react-router` directory.
For example:

```txt
- .react-router/
  - types/
    - app/
      - routes/
        - +types.product-details.ts
- app/
  - routes/
    - product-details.tsx
```

The path within `.react-router/types` purposefully mirrors the path to the correspond route module.
By setting things up like this, we can use `tsconfig.json`'s [rootDirs](https://www.typescriptlang.org/tsconfig/#rootDirs) option to let you conveniently import from the typegen file as if it was a sibling:

```ts
// app/routes/product-details.tsx
import * as Types from "./+types.product-details";
```

TypeScript will even give you import autocompletion for the typegen file and the `+` prefix helps to distinguish it as a special file.

> Big thanks to Svelte Kit for showing us that `rootDirs` trick!

Reviewer 3: Anything we can link to here?

### TypeScript plugin

> Typegen gets a bad rap because its a hassle when typegen'd files are out-of-date.
> Most typegen solutions require you to then rerun a script to update the typegen'd files.

Reviewer 3: Some suggestions on rewording. This is going to be the main criticism I believe, so I think a slightly less colloquial. You definitely don't have to go with this, I just think "bad rap" could potentially come across as dismissive, whereas "receive criticism" shows that we accept the negative feedback as legitimate.

```suggestion
Typegen solutions often receive criticism due to typegen'd files becoming out of sync during development.
This happens because many typegen solutions require you to then rerun a script to update the typegen'd files.
```

# Type inference

Date: 2024-09-20

Status: accepted

> Supercedes [#0003](./0003-infer-types-for-useloaderdata-and-useactiondata-from-loader-and-action-via-generics.md)

Reviewer: ```suggestion
Supersedes [#0003](./0003-infer-types-for-useloaderdata-and-useactiondata-from-loader-and-action-via-generics.md)
```

  // ...
}
```

We'll keep those hooks around for backwards compatibility, but eventually the aim is to deprecate and remove them.
We can design new, typesafe alternatives for any edge cases.

### Typegen

While React Router will default to programmatic routing, it can easily be configured for file-based routing.
That means that sometimes route URLs will only be represented as file paths.
Unfortunately, TypeScript cannot use the filesystem as part of its type inference nor type checking.
The only tenable way to infer types based on file paths is through code generation.

We _could_ have typegen just for file-based routing, but then we'd need to maintain a separate code path for type inference in programmatic routing.
To keep things simple, React Router treats any value returned by `routes.ts` the same; it will not make assumptions about _how_ those routes were constructed and will run typegen in all cases.

To that end, React Router will generate types for each route module into a special, gitignored `.react-router` directory.
For example:

```txt
- .react-router/
  - types/
    - app/
      - routes/
        - +types.product.ts
- app/
  - routes/
    - product.tsx
```

> The path within `.react-router/types` purposefully mirrors the path to the correspond route module.

Reviewer: ```suggestion
The path within `.react-router/types` purposefully mirrors the path to the corresponding route module.
```

- app/
  - routes/
    - product.tsx
```

The path within `.react-router/types` purposefully mirrors the path to the corresponding route module.
By setting things up like this, we can use `tsconfig.json`'s [rootDirs](https://www.typescriptlang.org/tsconfig/#rootDirs) option to let you conveniently import from the typegen file as if it was a sibling:

```ts
// app/routes/product.tsx
import { LoaderArgs, DefaultProps } from "./+types.product";
```

TypeScript will even give you import autocompletion for the typegen file and the `+` prefix helps to distinguish it as a special file.
Big thanks to Svelte Kit for showing us that [`rootDirs` trick](https://svelte.dev/blog/zero-config-type-safety#virtual-files)!

### TypeScript plugin

Typegen solutions often receive criticism due to typegen'd files becoming out of sync during development.
This happens because many typegen solutions require you to then rerun a script to update the typegen'd files.

Instead, our typegen will automatically run within a TypeScript plugin.
That means you should never need to manually run a typegen command during development.
It also means that you don't need to run our dev server for typegen to take effect.

> The only requirement is that your editor is open.

Reviewer 4: If it requires the editor to be open, how can we run type checking in CI?
